Source/WebDriver/WebDriverService.cpp

```diff
@@ -407,11 +413,99 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
     return true;
 }
 
-void WebDriverService::handleMessage(WebSocketMessageHandler::Message&&, Function<void (WebSocketMessageHandler::Message&&)>&&)
+void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message, Function<void (WebSocketMessageHandler::Message&&)>&& completionHandler)
 {
     // https://w3c.github.io/webdriver-bidi/#handle-an-incoming-message
-    // TODO Implement message handling for actual BiDi commands
-    // https://bugs.webkit.org/show_bug.cgi?id=280501
+
+    if (!message.connection) {
+        WTFLogAlways("Incoming message without attached connection. Ignoring message.");
+        completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::UnknownError, std::nullopt));
+        return;
+    }
+
+    auto connection = message.connection;
+    auto session = m_bidiServer.session(connection);
+    if (!session) {
+        if (!m_bidiServer.isStaticConnection(connection)) {
+            WTFLogAlways("Unknown connection. Ignoring message.");
+            completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidSessionID, connection));
+            return;
+        }
+    }
+    // 6.6 If session is null and command is not a static command, then send an error response given connection, command id, and invalid session id, and return.
+    // FIXME support checking static vs non-static methods https://bugs.webkit.org/show_bug.cgi?id=281721
+
+    auto parsedMessageValue = JSON::Value::parseJSON(String::fromUTF8(message.payload.data()));
+    if (!parsedMessageValue) {
+        WTFLogAlways("WebDriver handle Message: Failed to parse incoming message");
+        completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidArgument, message.connection));
+        return;
+    }
+
+    if (session && m_session && session->id() != m_session->id()) {
+        WTFLogAlways("Not an active session. Ignoring message.");
+        return;
+    }
+
+    BidiCommandHandler handler;
+    unsigned id = 0;
+    RefPtr<JSON::Object> parameters;
+    if (!findBidiCommand(parsedMessageValue, &handler, id, parameters)) {
+        WTFLogAlways("Failed to find appropriate BiDi command");
+        std::optional<int> commandId;
+        if (auto parsedMessageObject = parsedMessageValue->asObject()) {
+            auto parsedCommandId = parsedMessageObject->getInteger("id"_s);
+            if (parsedCommandId && *parsedCommandId >= 0)
+                commandId = parsedCommandId;
+        }
+
+        auto errorCode = CommandResult::ErrorCode::UnknownCommand;
+        auto errorReply = WebSocketMessageHandler::Message::fail(errorCode, connection, { "Command not supported"_s }, commandId);
+        completionHandler(WTFMove(errorReply));
+        return;
+    }
+
+    ((*this).*handler)(id, WTFMove(parameters), [completionHandler = WTFMove(completionHandler), message](std::optional<WebSocketMessageHandler::Message> resultMessage) {
```

Comment:
Why is resultMessage optional? Don't we always have to send a response?

Reply:
Indeed. The WebDriver classic handlers return a `CommandResult&&`. I've updated the WebDriver BiDi ones to also move a message back to the caller.